activerecord/lib/active_record/railtie.rb

```diff
@@ -319,9 +319,26 @@ class Railtie < Rails::Railtie # :nodoc:
       end
     end
 
-    initializer "active_record.set_signed_id_verifier_secret" do
-      ActiveSupport.on_load(:active_record) do
-        self.signed_id_verifier_secret ||= -> { Rails.application.key_generator.generate_key("active_record/signed_id") }
+    initializer "active_record.set_signed_id_verifier" do
+      config.after_initialize do |app|
+        ActiveSupport.on_load(:active_record) do
+          secret = self.signed_id_verifier_secret
+          secret_generator = if secret
+            ActiveRecord.deprecator.warn(<<~MSG)
+              ActiveRecord::Base.signed_id_verifier_secret is deprecated and will be removed in Rails ?.?. The secret will be derived from 'secret_key_base' instead.
```

Comment:
❓ What is the Rails version that we can remove `signed_id_verifier_secret`?

Reply:
I moved deprecation message into `ActiveRecord::SignedId`. I believe the rest of the logic should stay here in `railtie.rb ` file. wdyt?

Reply:
> signed_id_verifier_secret is also a class_attribute, which is unfortunate. That means existing apps could have, for example, set Post.signed_id_verifier_secret and Comment.signed_id_verifier_secret to different values, and Post.signed_id_verifier and Comment.signed_id_verifier each have to use those values, respectively.

We still don't support this one. I'm not sure what should we do in this case? It's very complicated to handle.

Reply:
I think I have an idea on how to approach this.  

If a `secret` is set for `ActiveRecord::Base`, we can use `message_verifiers` in `:generate_and_verify` mode. However, models that define their own `secret` should use a custom `signed_id_verifier`.  

In other words, there should be a global verifier initialized by `MessageVerifiers`, while models with a different `secret` can have their own custom verifiers.

Reply:
Yes, I touched on that a bit in https://github.com/rails/rails/pull/54422#discussion_r1943343277 :smiley::

> I have not fully sketched this out, but I would try giving `ActiveRecord::Base` access to `Rails.application.message_verifiers`. In other words, set something like `ActiveRecord::Base.message_verifiers` in an initializer, and use that instead of `MessageVerifier.new` in this method.

So instead of setting `ActiveRecord::Base.global_signed_id_verifier = app.message_verifier("active_record/signed_id")` in an initializer, we would set `ActiveRecord::Base.message_verifiers = app.message_verifiers`.

Then, in the `signed_id_verifier` method, we would return `ActiveRecord::Base.message_verifiers["active_record/signed_id"]` if `signed_id_verifier_secret` was not set.

Reply:
Ah yeah! 😃
I'll push the required changes for it.

Reply:
I'm using `class_attribute` for the `_signed_id_verifier` attribute, which introduces a slight change from the previous approach. Previously, we used `@signed_id_verifier`.  

I assume that to use `ActiveRecord` outside of Rails, `signed_id_verifier` should be explicitly set instead of relying on `signed_id_verifier_secret`. Setting `signed_id_verifier` for all models didn’t seem like a practical solution, so I opted to use `class_attribute` instead.